File: src/app/api/fetch_locations.rs
use std::sync::Arc;

use crate::{
    domain::{models::Location, use_cases},
    infra::LocationRepository,
};
use axum::{extract::State, http::StatusCode, Json};
use serde::Serialize;

#[derive(Serialize)]
pub struct Response {
    location: String,
}

impl From<Location> for Response {
    fn from(location: Location) -> Self {
        Self {
            location: location.into(),
        }
    }
}

type Error = (StatusCode, &'static str);

pub async fn fetch_locations(
    State(repo): State<Arc<dyn LocationRepository>>,
) -> Result<Json<Vec<Response>>, Error> {
    match use_cases::fetch_locations(repo).await {
        Ok(locations) => Ok(locations
            .into_iter()
            .map(|location| location.into())
            .collect::<Vec<Response>>()
            .into()),
        Err(use_cases::fetch_locations::Error::Unknown(str)) => {
            Err((StatusCode::INTERNAL_SERVER_ERROR, str))
        }
    }
}

File: src/app/api/fetchLocations.ts
import { Context } from "hono";
import * as lib from "@domain";
import { LocationRepository } from "@infra";
import { Result } from "@utils";

export async function fetchLocations(
    c: Context,
    locationRepo: LocationRepository
) {
    return Result.match(await lib.fetchLocations(locationRepo), {
        onOk: (locations) => c.json(locations.map(location => ({
            location: lib.Location.toString(location)
        }))),
        onErr: () => c.json([], 500)
    });
}
